### PR TITLE
refactor(plugins): switched from `require` to `await import()` when loading plugins

### PR DESCRIPTION
**Pull request by travi**
to isolate the conversion ahead of #2543

### CODE REVIEW

**Comment by travi** [comment]
this obviously still needs clean up, but hoping to get some feedback. i'm not 100% i have my head around all that needs to change here, but assuming we change from `require` to `await import()`, the majority of the work should be handling the additional promises that are introduced. i think i have most of that handled, but may have missed some.

however, the part that i did not anticipate is that the result does not appear to satisfy [`_.isPlainObject()`](https://lodash.com/docs/3.10.1#isPlainObject) the same way that loading with `require` did. it does not appear to be an issue with the added `default` property either.

logging one of the example plugins from the tests shows the following, which i would expect to both satisfy `_.isPlainObject` (at least because of the `null` prototype) and satisfy the expected shape if it were to make it far enough through the logic
```
[Module: null prototype] {
  analyzeCommits: [AsyncFunction: analyzeCommits],
  default: { analyzeCommits: [AsyncFunction: analyzeCommits] }
}
```
but it appears that the loaded module has some detail that i havent yet figured out that prevents it from satisfying `_.isPlainObject()`

**Comment by travi** [comment]
this also got me thinking about how this situation of loading plugins potentially gets more complicated as we progress further. currently, only cjs plugins are expected. once we move to esm, we will want to support loading plugins that are also esm. loading with `await import()` likely normalizes that situation for us, but we will want to confirm it once we are past this part.

this partially comes to mind because i did consider if it could be worth considering using [`module.createRequire`](https://nodejs.org/api/module.html#modulecreaterequirefilename) to avoid all of the complexity, but that would prevent supporting esm plugins. i dont think that even makes sense as an intermediate step since we will want to convert official plugins to esm as soon as converting core to esm unblocks converting the plugins.

File: lib/plugins/index.js [hunk 1]
```diff
@@ -6,36 +6,41 @@ const {validatePlugin, validateStep, loadPlugin, parseConfig} = require('./utils
...
-                });
-                plugins[type] = [...(plugins[type] || []), [func, config]];
-              }
-            });
-          } else {
-            errors.push(getError('EPLUGINSCONF', {plugin}));
-          }
+        console.log({plugins: options.plugins})
+        console.dir(plugin, {depth: null})
+        console.log(isPlainObject(plugin))
```
Comment (gr2m) [inline lib/plugins/index.js:22]: just a reminder to remove the debug logs when the PR is ready :)

File: lib/plugins/index.js [hunk 2]
```diff
@@ -6,36 +6,41 @@ const {validatePlugin, validateStep, loadPlugin, parseConfig} = require('./utils
...
   let {options, logger} = context;
   const errors = [];
 
   const plugins = options.plugins
-    ? castArray(options.plugins).reduce((plugins, plugin) => {
-        if (validatePlugin(plugin)) {
-          const [name, config] = parseConfig(plugin);
-          plugin = isString(name) ? loadPlugin(context, name, pluginsPath) : name;
+    ? await castArray(options.plugins).reduce(async (plugins, plugin) => {
+      const plugins2 = await plugins;
```
Comment (gr2m) [inline lib/plugins/index.js:15]: I wonder if there is a better name for `plugins2`? Maybe `normalizedPlugins` or `awaitedPlugins`?

**Review by gr2m — COMMENTED** [review]
changes look good to me thus far

File: lib/plugins/index.js (hunk 2 above)
Comment (travi) [reply lib/plugins/index.js:15]: yep, this is definitely part of the clean up that is necessary once this is working :)

File: lib/plugins/index.js (hunk 1 above)
Comment (travi) [reply lib/plugins/index.js:22]: yep 👍🏼

**Comment by travi** [comment]
i was able to resolve the previous issues by fixing the paren wrapping that i broke (this was the main problem) and using the `.default` property of the imported module (which i had tried previously, but wasnt working fully because of the former problem

**Comment by travi** [comment]
however, `xo` now reports a lint issue for the dynamic import expression:
```
lib/plugins/utils.js:52:43
  ✖  52:43  import() expressions are not supported yet.  node/no-unsupported-features/es-syntax
```

however, all tests, including integration tests pass in our supported node range from my checking. all scenarios within the dynamic import expression seem to be covered by tests as well.

based on https://github.com/mysticatea/eslint-plugin-node/issues/250, it sounds like the rule does not report correctly for this situation and is unlikely to be fixed soon. i hate to disable a rule like this and lose protection from other scenarios, but i'm unsure of a better approach for now. i'm open to suggestions

**Comment by travi** [comment]
i went ahead and disabled the rule, since it seems like the best option, but open to alternative suggestions

**Comment by gr2m** [comment]
> i hate to disable a rule like this and lose protection from other scenarios, but i'm unsure of a better approach for now. i'm open to suggestions

+1 for disabling. We can create a follow up issue to review once we concluded the ESM transition?

Great work @travi 👏🏼

**Pull request merged**